pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -0,0 +1,552 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package noderoute
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	mcv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	mcclientset "antrea.io/antrea/multicluster/pkg/client/clientset/versioned"
+	mcinformers "antrea.io/antrea/multicluster/pkg/client/informers/externalversions/multicluster/v1alpha1"
+	mclisters "antrea.io/antrea/multicluster/pkg/client/listers/multicluster/v1alpha1"
+	"antrea.io/antrea/pkg/agent/config"
+	"antrea.io/antrea/pkg/agent/interfacestore"
+	"antrea.io/antrea/pkg/agent/openflow"
+	"antrea.io/antrea/pkg/ovs/ovsconfig"
+)
+
+const (
+	controllerName = "AntreaAgentMCRouteController"
+
+	// Set resyncPeriod to 0 to disable resyncing
+	resyncPeriod = 0 * time.Second
+	// How long to wait before retrying the processing of a resource change
+	minRetryDelay = 2 * time.Second
+	maxRetryDelay = 120 * time.Second
+
+	// Default number of workers processing a resource change
+	defaultWorkers = 1
+
+	gatewayKind  = "Gateway"
+	ciImportKind = "ClusterInfoImport"
+)
+
+type key struct {
+	kind string
+	name string
+}
+
+type ciImportInfo struct {
+	localGWName string
+	ciImport    mcv1alpha1.ClusterInfoImport
+}
+
+// MCRouteController watches Gateway and ClusterInfoImport events.
+// It is responsible for setting up necessary Openflow entries for multi-cluster
+// traffic on a Gateway or a regular Node.
+type MCRouteController struct {
+	mcClient             mcclientset.Interface
+	ovsBridgeClient      ovsconfig.OVSBridgeClient
+	ofClient             openflow.Client
+	interfaceStore       interfacestore.InterfaceStore
+	nodeConfig           *config.NodeConfig
+	gwInformer           mcinformers.GatewayInformer
+	gwLister             mclisters.GatewayLister
+	gwListerSynced       cache.InformerSynced
+	ciImportInformer     mcinformers.ClusterInfoImportInformer
+	ciImportLister       mclisters.ClusterInfoImportLister
+	ciImportListerSynced cache.InformerSynced
+	queue                workqueue.RateLimitingInterface
+	// installedCIImports is for saving ClusterInfos which have been processed
+	// in MCRouteController.
+	installedCIImports cache.Indexer
+	// Need to use mutex to protect 'installedGWName' if we plan to change
+	// the number of 'defaultWorkers' to run multiple go rountines to handle
+	// events.
+	installedGWName string
+	namespace       string
```

Comment:
Done, comment is added, it's the Namespace where Antrea multi-cluster controller running, by default, it will be deployed to `kube-system` which is the same as Antrea-agent, so I used `env.GetPodNamespace()`, I am wondering if it's Ok to do like this and update the doc to remind user about that Antrea and Antrea MC member controller should be running in the same Namespace, or I should add a config in antrea-agent.conf, any suggestion? thanks.

Reply:
added a new config for multicluster.